src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,695 @@
+                }
+
+                if (array.Length - index < _queue._size)
+                {
+                    throw new ArgumentException(SR.Argument_InvalidOffLen);
+                }
+
+                try
+                {
+                    Array.Copy(_queue._nodes, 0, array, index, _queue._size);
+                }
+                catch (ArrayTypeMismatchException)
+                {
+                    throw new ArgumentException(SR.Argument_InvalidArrayType, nameof(array));
+                }
+            }
+
+            public struct Enumerator : IEnumerator<(TElement element, TPriority priority)>, IEnumerator
+            {
+                private readonly PriorityQueue<TElement, TPriority> _queue;
+                private readonly int _version;
+
+                private int _index;
+                private (TElement element, TPriority priority)? _currentElement;
+
+                private const int FirstCallToEnumerator = -2;
+                private const int EndOfEnumeration = -1;
+
+                internal Enumerator(PriorityQueue<TElement, TPriority> queue)
+                {
+                    _queue = queue;
+                    _version = queue._version;
+                    _index = FirstCallToEnumerator;
+                    _currentElement = default;
+                }
+
+                public void Dispose()
+                {
+                    _index = EndOfEnumeration;
+                }
+
+                public bool MoveNext()
+                {
+                    bool advancedEnumerator;
+
+                    if (_version != _queue._version)
+                    {
+                        throw new InvalidOperationException(SR.InvalidOperation_EnumFailedVersion);
+                    }
+
+                    if (_index == FirstCallToEnumerator)
+                    {
+                        _index = 0;
+                        advancedEnumerator = (_queue._size > 0);
+
+                        if (advancedEnumerator)
+                        {
+                            _currentElement = _queue._nodes[_index];
+                        }
+                        else
+                        {
+                            _index = EndOfEnumeration;
+                        }
+
+                        return advancedEnumerator;
+                    }
+
+                    if (_index == EndOfEnumeration)
+                    {
+                        return false;
+                    }
+
+                    _index++;
+                    advancedEnumerator = (_index < _queue._size);
+
+                    if (advancedEnumerator)
+                    {
+                        _currentElement = _queue._nodes[_index];
+                    }
+                    else
+                    {
+                        _currentElement = default;
+                    }
+
+                    return advancedEnumerator;
+                }
```

Comment:
This implementation looks more complicated than it needs to be.  Can you look at the Enumerator for `List<T>` or `Queue<T>` can effectively copy one of those?

Reply:
It's actually adapted `Stack<T>` 🙈

Reply:
The `Stack<T>` one looks a lot simpler, as does `Queue<T>`. It should not be necessary to introduce any local variables.

How about this
```c#
            public struct Enumerator : IEnumerator<(TElement element, TPriority priority)>
            {
                private readonly PriorityQueue<TElement, TPriority> _queue;
                private readonly int _version;

                private int _index;   // -1 = not started, -2 = ended/disposed
                private (TElement element, TPriority priority)? _currentElement;

                internal Enumerator(PriorityQueue<TElement, TPriority> queue)
                {
                    _queue = queue;
                    _version = queue._version;
                    _index = -1;
                    _currentElement = default;
                }

                public void Dispose()
                {
                    _index = -2;
                    _currentElement = default;
                }

                public bool MoveNext()
                {
                    if (_version != _queue._version)
                       throw new InvalidOperationException(SR.InvalidOperation_EnumFailedVersion);

                    if (_index == -2) // ended
                        return false;

                    _index++;

                    if (_index < _queue._size)
                    {
                        _currentElement = _queue._nodes[_index];
                        return true;
                    }

                    _currentElement == default;
                    _index = -2;
                    return true;
                }

                public (TElement element, TPriority priority) Current =>
                {
                    get
                    {
                        if (_index < 0)
                            ThrowEnumerationNotStartedOrEnded();
                        return _currentElement;
                    }
                }
     
                private void ThrowEnumerationNotStartedOrEnded() =>
                    throw new InvalidOperationException(_index == -1 ? SR.InvalidOperation_EnumNotStarted : SR.InvalidOperation_EnumEnded);
                }
     
                object? System.Collections.IEnumerator.Current => Current;
     
                void IEnumerator.Reset()
                {
                    if (_version != _queue._version)
                        throw new InvalidOperationException(SR.InvalidOperation_EnumFailedVersion);
                    _index = -1;
                    _currentElement = default;
                }
            }
```

Reply:
I don't feel strongly about having the consts or not -- I think it's self explanatory enough to not need th extra IL of consts - but fine either way here (unlike the growth factor, they're used in several places, and have arbitrary values)

Reply:
I didn't bother to make Dispose() neutralize the enumerator as that doesn't seem like what we normally do

Reply:
If enumeration is perf-critical, you can use the tricks in `List<T>`
1. Grab fields into locals so they're read once in a method
2. Break into MoveNext and MoveNextRare.
3. Exatract throw new InvalidOperationException(SR.InvalidOperation_EnumFailedVersion); into a ThrowXXX method asit makes less IL.